pool/paymentmgr.go

```diff
@@ -0,0 +1,538 @@
+	now := nanoToBigEndianBytes(time.Now().UnixNano())
+	lastPaymentCreatedOn := pm.fetchLastPaymentCreatedOn()
+	shares, err := PPSEligibleShares(pm.cfg.DB, nanoToBigEndianBytes(int64(lastPaymentCreatedOn)), now)
+	if err != nil {
+		return nil, err
+	}
+	if len(shares) == 0 {
+		return make(map[string]*big.Rat), nil
+	}
+	percentages, err := sharePercentages(shares)
+	if err != nil {
+		return nil, err
+	}
+	return percentages, nil
+}
+
+// PPLNSSharePercentages calculates the current mining reward percentages due pool
+// accounts based on work performed measured by the PPLNS payment scheme.
+func (pm *PaymentMgr) PPLNSSharePercentages() (map[string]*big.Rat, error) {
+	now := time.Now()
+	min := now.Add(-(time.Second * time.Duration(pm.cfg.LastNPeriod)))
+	minNano := nanoToBigEndianBytes(min.UnixNano())
+	shares, err := PPLNSEligibleShares(pm.cfg.DB, minNano)
+	if err != nil {
+		return nil, err
+	}
+	if len(shares) == 0 {
+		return make(map[string]*big.Rat), nil
+	}
+
+	// Deduct pool fees and calculate the payment due each participating
+	// account.
+	percentages, err := sharePercentages(shares)
+	if err != nil {
+		return nil, err
+	}
+	return percentages, nil
+}
+
+// PayPerShare generates a payment bundle comprised of payments to all
+// participating accounts. Payments are calculated based on work contributed
+// to the pool since the last payment batch.
+func (pm *PaymentMgr) payPerShare(coinbase dcrutil.Amount, height uint32) error {
+	now := time.Now()
+	percentages, err := pm.PPSSharePercentages()
+	if err != nil {
+		return err
+	}
+	coinbaseMaturity := pm.cfg.ActiveNet.CoinbaseMaturity
+	var estMaturity uint32
+	if coinbaseMaturity == 0 {
+		// Allow immediately mature payments for testing purposes.
+		estMaturity = height
+	}
+	if coinbaseMaturity > 0 {
+		estMaturity = height + uint32(coinbaseMaturity)
+	}
+	payments, err := CalculatePayments(percentages, coinbase, pm.cfg.PoolFee, height, estMaturity)
+	if err != nil {
+		return err
+	}
+	for _, payment := range payments {
+		err := payment.Create(pm.cfg.DB)
+		if err != nil {
+			return err
+		}
+	}
+	lastPaymentCreatedOn := uint64(payments[len(payments)-1].CreatedOn)
+	pm.setLastPaymentCreatedOn(lastPaymentCreatedOn)
+	err = pm.cfg.DB.Update(func(tx *bolt.Tx) error {
+		// Update the last payment created on time and prune invalidated shares.
+		err := pm.persistLastPaymentCreatedOn(tx)
+		if err != nil {
+			return err
+		}
+		return pruneShares(tx, now.UnixNano())
+	})
+	return err
+}
+
+// payPerLastNShares generates a payment bundle comprised of payments to all
+// participating accounts within the lastNPeriod of the pool.
+func (pm *PaymentMgr) payPerLastNShares(coinbase dcrutil.Amount, height uint32) error {
+	percentages, err := pm.PPLNSSharePercentages()
+	if err != nil {
+		return err
+	}
+	var estMaturity uint32
+	coinbaseMaturity := pm.cfg.ActiveNet.CoinbaseMaturity
+	if coinbaseMaturity == 0 {
```

Comment:
Coinbase maturity is never set to 0 in any of the dcrd netparam files

Reply:
Yes, and it is never going to be zero in production. This just allows running tests with a mature coinbase once a block it is mined.